File: src/tropicam/time_surface.py
"""The time surface: T(x, y) <- max(T(x, y), t).

That update rule is tropical (max-plus) addition. The whole project follows
from taking it literally, so this module is deliberately small and exact.

Design decisions locked in by the master doc (section 7):

* **Pre-allocated flat contiguous array**, shape (polarities, H, W), row-major.
  Neighbourhood access is what buys locality; fitting in L2 is not the goal.

* **uint32 timestamps with wrapping unsigned subtraction** for age. `now - T`
  is correct mod 2**32 as long as the true age is below ~71.6 minutes at
  microsecond resolution.

* **No rolling epoch.** Rebasing timestamps to keep them small forces a
  periodic O(N) stop-the-world sweep over the entire surface, and that pause
  lands squarely in the p99 latency plot we care about. Wrapping subtraction
  gives the same memory saving with no sweep and no spike.

* **Scatter-max, not last-write-wins.** See `scatter_max` below; this is the
  correctness footgun the doc calls out by name.
"""

from __future__ import annotations

import numpy as np

#: Timestamps wrap at this value; ages above it are ambiguous.
WRAP_US = 1 << 32

#: ~71.58 minutes. Ages beyond this cannot be represented unambiguously.
MAX_UNAMBIGUOUS_AGE_US = WRAP_US - 1


def scatter_max(surface_flat: np.ndarray, flat_idx: np.ndarray, ts: np.ndarray,
                method: str = "sort") -> None:
    """In-place ``surface_flat[i] = max(surface_flat[i], t)`` over duplicates.

    The obvious spelling is wrong::

        T[ys, xs] = np.maximum(T[ys, xs], ts)   # WRONG

    Fancy-index assignment is **last-write-wins** when `flat_idx` repeats
    within a batch, so a later-arriving *older* timestamp silently overwrites a
    newer one. It is not a max. A clean translating bar fires each pixel
    exactly once, so the synthetic oracle will NOT catch this -- hence the
    explicit duplicate-pixel test in the suite.

    Two correct implementations, kept side by side because the choice matters
    once the hot loop gets ported:

    ``"ufunc"``
        ``np.maximum.at`` -- unbuffered, unambiguously correct, slow.
    ``"sort"``
        Sort by (index, timestamp), keep the last row per index, then assign.
        Indices are unique after the reduction, so plain fancy assignment is
        safe. Usually several times faster on realistic batch sizes.
    """
    if flat_idx.size == 0:
        return

    if method == "ufunc":
        np.maximum.at(surface_flat, flat_idx, ts)
        return

    if method != "sort":
        raise ValueError(f"unknown scatter_max method: {method!r}")

    # lexsort orders by the LAST key first, so this is (idx asc, then t asc).
    order = np.lexsort((ts, flat_idx))
    idx_sorted = flat_idx[order]
    ts_sorted = ts[order]

    # Keep the final row of each run of equal indices == the max timestamp.
    keep = np.empty(idx_sorted.shape, dtype=bool)
    keep[-1] = True
    np.not_equal(idx_sorted[1:], idx_sorted[:-1], out=keep[:-1])

    uniq_idx = idx_sorted[keep]
    uniq_ts = ts_sorted[keep]

    # Indices are unique now, so this is a genuine element-wise max.
    surface_flat[uniq_idx] = np.maximum(surface_flat[uniq_idx], uniq_ts)


class TimeSurface:
    """Per-pixel most-recent-event timestamps, one plane per polarity.

    Polarity planes are kept separate by default (doc section 7.3): merging
    them discards the sign of the brightness change, and the decision changes
    the memory footprint enough to matter (~7.4 MB at 1280x720, 2 planes,
    32-bit).
    """

    def __init__(self, height: int, width: int, n_polarities: int = 2,
                 scatter_method: str = "sort"):
        if height <= 0 or width <= 0:
            raise ValueError("height and width must be positive")
        self.height = int(height)
        self.width = int(width)
        self.n_polarities = int(n_polarities)
        self.scatter_method = scatter_method

        # Contiguous, pre-allocated, never reallocated.
        self._t = np.zeros((self.n_polarities, self.height, self.width),
                           dtype=np.uint32)
        self._flat = self._t.reshape(-1)
        self._plane_stride = self.height * self.width

        #: Timestamp of the most recent event consumed, for age queries.
        self.now = np.uint32(0)

    # -- ingest ---------------------------------------------------------

    def update(self, events: np.ndarray) -> None:
        """Consume a batch of events. Order within the batch does not matter."""
        if events.size == 0:
            return
        x = events["x"].astype(np.intp)
        y = events["y"].astype(np.intp)
        p = events["p"].astype(np.intp)
        t = events["t"]

        flat_idx = (p * self._plane_stride) + (y * self.width) + x
        scatter_max(self._flat, flat_idx, t, method=self.scatter_method)

        batch_max = t.max()
        if batch_max > self.now:
            self.now = np.uint32(batch_max)

    def reset(self) -> None:
        self._t.fill(0)
        self.now = np.uint32(0)

    # -- read -----------------------------------------------------------

    @property
    def timestamps(self) -> np.ndarray:
        """(polarities, H, W) view of raw timestamps. Do not mutate."""
        return self._t

    def plane(self, polarity: int) -> np.ndarray:
        """(H, W) view of one polarity plane."""
        return self._t[polarity]

    def merged(self) -> np.ndarray:
        """(H, W) polarity-agnostic surface -- the max across planes.

        Tropical addition across planes, which is the only merge that respects
        the semiring. Averaging the planes would be exactly the clock-time
        mistake.
        """
        return self._t.max(axis=0)

    def age(self, now: int | None = None) -> np.ndarray:
        """Age of each pixel in microseconds, via wrapping unsigned subtraction.

        Valid only where the true age is below `MAX_UNAMBIGUOUS_AGE_US`; older
        pixels alias. Quiet regions therefore report garbage ages, which is
        acceptable because they carry no motion evidence -- but do not treat a
        large age as meaningful without checking `has_fired`.
        """
        ref = self.now if now is None else np.uint32(now)
        return ref - self._t  # uint32 arithmetic wraps; this is intentional.

    def has_fired(self) -> np.ndarray:
        """Boolean mask of pixels that have received at least one event.

        Timestamp 0 is the sentinel for "never fired". A genuine event at
        t == 0 is indistinguishable from it; the synthetic oracle therefore
        starts streams at t > 0.
        """
        return self._t != 0

    # -- introspection --------------------------------------------------

    @property
    def nbytes(self) -> int:
        return self._t.nbytes

    def __repr__(self) -> str:
        return (f"TimeSurface({self.height}x{self.width}, "
                f"{self.n_polarities} planes, {self.nbytes / 1e6:.1f} MB, "
                f"now={int(self.now)}us)")

File: crates/tropicam-core/src/time_surface.rs
//! `T(x, y) <- max(T(x, y), t)` -- tropical addition, one event at a time.
//!
//! This is where the native port actually earns its keep, and the reason is
//! worth stating precisely because it is a project claim, not a micro-
//! optimisation.
//!
//! In NumPy, a correct scatter-max over a batch needs either `np.maximum.at`
//! (unbuffered, slow) or a lexsort-and-reduce -- **O(n log n)** in the batch
//! size, because vectorised fancy-index assignment resolves duplicate pixels
//! by last write rather than by max.
//!
//! In Rust the same operation is a load, a compare, and a conditional store.
//! No sort, no allocation, no batching required:
//!
//! ```text
//! let slot = &mut self.t[idx];
//! if ev.t > *slot { *slot = ev.t; }
//! ```
//!
//! That is **strictly constant work per event**, independent of batch size,
//! sensor resolution, and stream history -- which is the O(1)-per-event bound
//! the project scope commits to. The NumPy path cannot make that claim at
//! all; it was always a prototype standing in for this.

/// Timestamps wrap at 2^32 us; ages above this alias (~71.6 minutes).
pub const MAX_UNAMBIGUOUS_AGE_US: u32 = u32::MAX;

/// Per-pixel most-recent-event timestamps, one plane per polarity.
///
/// Pre-allocated, flat, contiguous, row-major. Never reallocated after
/// construction, so ingest performs no allocation whatsoever.
pub struct TimeSurface {
    t: Vec<u32>,
    width: usize,
    height: usize,
    planes: usize,
    plane_stride: usize,
    now: u32,
}

/// Outcome of ingesting a batch: enough to verify the work bound in tests.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct IngestStats {
    /// Events consumed.
    pub consumed: u64,
    /// Events that actually advanced a pixel's timestamp.
    pub updated: u64,
    /// Events dropped for being out of bounds.
    pub out_of_bounds: u64,
}

impl TimeSurface {
    pub fn new(height: usize, width: usize, planes: usize) -> Self {
        assert!(height > 0 && width > 0 && planes > 0);
        let plane_stride = height * width;
        Self {
            t: vec![0u32; plane_stride * planes],
            width,
            height,
            planes,
            plane_stride,
            now: 0,
        }
    }

    #[inline(always)]
    fn index(&self, x: u16, y: u16, p: u8) -> Option<usize> {
        let (x, y, p) = (x as usize, y as usize, p as usize);
        if x >= self.width || y >= self.height || p >= self.planes {
            return None;
        }
        Some(p * self.plane_stride + y * self.width + x)
    }

    /// Consume one event. Constant work; no allocation, no sort, no branchy
    /// bookkeeping. Order of arrival within a batch does not matter.
    #[inline(always)]
    pub fn push(&mut self, ev: crate::Event) -> bool {
        let (x, y, t, p) = (ev.x(), ev.y(), ev.t(), ev.p());
        match self.index(x, y, p) {
            None => false,
            Some(idx) => {
                // SAFETY: index() bounds-checked against the allocation.
                let slot = unsafe { self.t.get_unchecked_mut(idx) };
                if t > *slot {
                    *slot = t;
                }
                if t > self.now {
                    self.now = t;
                }
                true
            }
        }
    }

    /// Consume a batch. Exactly `push` in a loop -- kept separate only so the
    /// FFI boundary is crossed once per batch instead of once per event.
    pub fn extend(&mut self, events: &[crate::Event]) -> IngestStats {
        let mut stats = IngestStats {
            consumed: events.len() as u64,
            ..Default::default()
        };
        for &ev in events {
            let t = ev.t();
            match self.index(ev.x(), ev.y(), ev.p()) {
                None => stats.out_of_bounds += 1,
                Some(idx) => {
                    let slot = unsafe { self.t.get_unchecked_mut(idx) };
                    if t > *slot {
                        *slot = t;
                        stats.updated += 1;
                    }
                    if t > self.now {
                        self.now = t;
                    }
                }
            }
        }
        stats
    }

    pub fn reset(&mut self) {
        self.t.fill(0);
        self.now = 0;
    }

    // -- read -----------------------------------------------------------

    #[inline]
    pub fn timestamps(&self) -> &[u32] {
        &self.t
    }

    pub fn plane(&self, p: usize) -> &[u32] {
        let start = p * self.plane_stride;
        &self.t[start..start + self.plane_stride]
    }

    /// Polarity-agnostic surface: the max across planes.
    ///
    /// Tropical addition, which is the only merge respecting the semiring.
    /// Averaging the planes would be the clock-time mistake in miniature.
    pub fn merged(&self) -> Vec<u32> {
        let mut out = self.plane(0).to_vec();
        for p in 1..self.planes {
            for (o, &v) in out.iter_mut().zip(self.plane(p)) {
                if v > *o {
                    *o = v;
                }
            }
        }
        out
    }

    /// Age per pixel via wrapping unsigned subtraction -- no rolling epoch,
    /// hence no periodic O(N) stop-the-world sweep in the p99 plot.
    pub fn age(&self, now: u32) -> Vec<u32> {
        self.t.iter().map(|&v| now.wrapping_sub(v)).collect()
    }

    #[inline]
    pub fn now(&self) -> u32 {
        self.now
    }
    #[inline]
    pub fn width(&self) -> usize {
        self.width
    }
    #[inline]
    pub fn height(&self) -> usize {
        self.height
    }
    #[inline]
    pub fn planes(&self) -> usize {
        self.planes
    }
    #[inline]
    pub fn nbytes(&self) -> usize {
        self.t.len() * core::mem::size_of::<u32>()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::Event;

    #[test]
    fn max_not_last_write_wins() {
        // The footgun, in Rust: oldest timestamp arrives last and must lose.
        let mut ts = TimeSurface::new(8, 8, 2);
        ts.extend(&[
            Event::new(3, 4, 500, 1),
            Event::new(3, 4, 900, 1),
            Event::new(3, 4, 100, 1),
        ]);
        assert_eq!(ts.plane(1)[4 * 8 + 3], 900);
    }

    #[test]
    fn order_independent() {
        let a = {
            let mut s = TimeSurface::new(16, 16, 2);
            s.extend(&[
                Event::new(1, 1, 10, 1),
                Event::new(1, 1, 90, 1),
                Event::new(2, 2, 50, 0),
            ]);
            s.timestamps().to_vec()
        };
        let b = {
            let mut s = TimeSurface::new(16, 16, 2);
            s.extend(&[
                Event::new(2, 2, 50, 0),
                Event::new(1, 1, 90, 1),
                Event::new(1, 1, 10, 1),
            ]);
            s.timestamps().to_vec()
        };
        assert_eq!(a, b);
    }

    #[test]
    fn wrapping_age_across_the_boundary() {
        let mut ts = TimeSurface::new(4, 4, 2);
        ts.extend(&[Event::new(0, 0, u32::MAX - 99, 1)]);
        // now has wrapped past zero; true age is 350us.
        let age = ts.age(250);
        assert_eq!(age[ts.plane_stride], 350);
    }

    #[test]
    fn polarity_planes_independent_and_merge_is_max() {
        let mut ts = TimeSurface::new(8, 8, 2);
        ts.extend(&[Event::new(2, 3, 400, 1), Event::new(2, 3, 800, 0)]);
        let i = 3 * 8 + 2;
        assert_eq!(ts.plane(1)[i], 400);
        assert_eq!(ts.plane(0)[i], 800);
        assert_eq!(ts.merged()[i], 800);
    }

    #[test]
    fn out_of_bounds_dropped_not_wrapped() {
        let mut ts = TimeSurface::new(8, 8, 2);
        let stats = ts.extend(&[Event::new(99, 0, 5, 1), Event::new(0, 99, 5, 1)]);
        assert_eq!(stats.out_of_bounds, 2);
        assert_eq!(ts.timestamps().iter().copied().max(), Some(0));
    }

    #[test]
    fn footprint_matches_budget() {
        // Doc section 7.3: ~7.4 MB at 1280x720, two 32-bit polarity planes.
        let ts = TimeSurface::new(720, 1280, 2);
        assert!(ts.nbytes() > 7_000_000 && ts.nbytes() < 7_500_000);
    }
}
